plugin/query_predicates.lua

```diff
@@ -45,29 +45,37 @@ query.add_directive('set-lang-from-mimetype!', function(match, _, bufnr, pred, m
   end
 end, true)
 
-local injection_aliases = {
-  ex = 'elixir',
-  pl = 'perl',
-  sh = 'bash',
-  uxn = 'uxntal',
-  ts = 'typescript',
-}
-
+--- Set injection.language metadata from node text
+---
+--- If text does not yield an installed parser name, treat as `filetype`
+--- and look up corresponding parser.
+---
+--- Custom aliases can be added via |vim.treesitter.language.register()|.
 ---@param match (TSNode|nil)[]
 ---@param _ string
 ---@param bufnr integer
 ---@param pred string[]
 ---@return boolean|nil
-query.add_directive('set-lang-from-info-string!', function(match, _, bufnr, pred, metadata)
-  local capture_id = pred[2]
-  local node = match[capture_id]
+query.add_directive('inject-language!', function(match, _, bufnr, pred, metadata)
+  local node = match[pred[2]]
   if not node then
     return
   end
 
-  local injection_alias = vim.treesitter.get_node_text(node, bufnr)
-  local filetype = vim.filetype.match({ filename = 'a.' .. injection_alias })
-  metadata['injection.language'] = filetype or injection_aliases[injection_alias] or injection_alias
+  local has_parser = function(lang)
+    return vim._ts_has_language(lang)
+      or #vim.api.nvim_get_runtime_file('parser/' .. lang .. '.*', false) > 0
```

Comment:
Hmm, that doesn't seem to work with non-language name annotations (like `tex` for `latex`).

Reply:
and it can't work (currently), since `vim.treesitter.language.add` unconditionally registers `lang` as a valid alias, completely defeating the purpose.

So we'll need a new API for such checks in core (which is a good idea anyway). Leaving this as-is for now, but with a todo note (as that should indeed be handled opaquely by the API).